src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -1142,21 +1150,35 @@ void ObjectSynchronizer::monitors_iterate(MonitorClosure* closure) {
 }
 
 static bool monitors_used_above_threshold(MonitorList* list) {
+  if (MonitorUsedDeflationThreshold == 0) {  // disabled case is easy
+    return false;
+  }
   // Start with ceiling based on a per-thread estimate:
   size_t ceiling = ObjectSynchronizer::in_use_list_ceiling();
   if (ceiling < list->max()) {
     // The max used by the system has exceeded the ceiling so use that:
     ceiling = list->max();
   }
-  if (ceiling == 0) {
+  size_t monitors_used = list->count();
+  if (monitors_used == 0) {  // empty list is easy
     return false;
   }
-  if (MonitorUsedDeflationThreshold > 0) {
-    size_t monitors_used = list->count();
-    size_t monitor_usage = (monitors_used * 100LL) / ceiling;
-    return int(monitor_usage) > MonitorUsedDeflationThreshold;
-  }
-  return false;
+  if (NoAsyncDeflationProgressMax != 0 &&
+      _no_progress_cnt >= NoAsyncDeflationProgressMax) {
+    float remainder = (100.0 - MonitorUsedDeflationThreshold) / 100.0;
+    size_t new_ceiling = ceiling + (ceiling * remainder) + 1;
+    size_t old_ceiling = ObjectSynchronizer::in_use_list_ceiling();
```

Comment:
You could capture:

size_t old_ceiling = ceiling;

back after line 1157, rather than re-reading it.

Reply:
Fixed.